optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,168 @@
+            "to `False` in the constructor of `QMCSampler`, "
+            "if this independent sampling is intended behavior.".format(
+                param_name, trial.number, self._independent_sampler.__class__.__name__
+            )
+        )
+
+    def _reset_qmc_engine(self, d: int) -> None:
+
+        # Lazy import because the `scipy.stats.qmc` is slow to import.
+        import scipy.stats.qmc
+
+        self._samples_count = 0  # The number of samples, taken from the engine.
+        scramble = False if self._seed is None else True
+        if self._qmc_type == "sobol":
+            self._qmc_engine = scipy.stats.qmc.Sobol(d, seed=self._seed, scramble=scramble)
+        elif self._qmc_type == "halton":
+            self._qmc_engine = scipy.stats.qmc.Halton(d, seed=self._seed, scramble=scramble)
+        elif self._qmc_type == "LHS":  # Latin Hypercube Sampling
+            self._qmc_engine = scipy.stats.qmc.Latin(d, seed=self._seed)
+        elif self._qmc_type == "OA-LHS":  # Orthogonal array-based Latin hypercube sampling
+            self._qmc_engine = scipy.stats.qmc.OrthogonalLatinHypercube(d, seed=self._seed)
+        else:
+            message = (
+                "The `qmc_type`, {}, is not a valid. "
+                "It must be one of sobol, halton, LHS, and OA-LHS."
+            )
+            raise ValueError(message)
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+
+        return self._independent_sampler.sample_independent(
+            study, trial, param_name, param_distribution
+        )
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> Dict[str, Any]:
+
+        if search_space == {}:
+            return {}
+
+        assert self._initial_search_space is not None
+
+        if self._qmc_engine is None:
+            n_initial_params = len(self._initial_search_space)
+            self._reset_qmc_engine(n_initial_params)
+
+        assert isinstance(self._qmc_engine, scipy.stats.qmc.QMCEngine)
+
+        qmc_id = self._find_qmc_id(study, trial)
+        forward_size = qmc_id - self._qmc_engine.num_generated  # `qmc_id` starts from 0.
+        self._qmc_engine.fast_forward(forward_size)
+        sample = self._qmc_engine.random(1)
+
+        trans = _SearchSpaceTransform(search_space)
+        sample = scipy.stats.qmc.scale(sample, trans.bounds[:, 0], trans.bounds[:, 1])
+        sample = trans.untransform(sample[0, :])
+
+        return sample
+
+    def _find_qmc_id(self, study: Study, trial: FrozenTrial) -> int:
+        # TODO(kstoneriv3): Following try-except block assumes that the block is
+        # an atomic transaction. # This ensures that each qmc_id is sampled at least once.
+        key_qmc_id = "{}_last_qmc_id".format(self._qmc_type)
```

Comment:
A nit. You can use f-string.
```suggestion
        key_qmc_id = f"{self._qmc_type}_last_qmc_id"
```